tests/middleware/test_base.py

```diff
@@ -206,3 +213,81 @@ async def homepage(request):
     client = test_client_factory(app)
     response = client.get("/")
     assert response.status_code == 200, response.content
+
+
+@pytest.mark.anyio
+async def test_background_tasks_client_disconnect() -> None:
+    # test for https://github.com/encode/starlette/issues/1438
+    container: List[str] = []
+
+    disconnected = anyio.Event()
+
+    async def slow_background() -> None:
+        await disconnected.wait()
+        container.append("called")
+
+    app: ASGIApp
+    app = PlainTextResponse("hi!", background=BackgroundTask(slow_background))
+
+    async def dispatch(
+        request: Request, call_next: Callable[[Request], Awaitable[Response]]
+    ) -> Response:
+        return await call_next(request)
+
+    app = BaseHTTPMiddleware(app, dispatch=dispatch)
+
+    app = BackgroundTaskMiddleware(app)
+
+    async def recv_gen() -> AsyncGenerator[Message, None]:
+        yield {"type": "http.request"}
+        disconnected.set()
+        yield {"type": "http.disconnect"}
+
+    async def send_gen() -> AsyncGenerator[None, Message]:
+        msg = yield
+        assert msg["type"] == "http.response.start"
+        await disconnected.wait()
+        raise AssertionError("Should not be called")  # pragma: no cover
```

Comment:
I would say that this sequence of events does not look natural to me, in terms of expressing what happens in #1438.

I believe the main event that should be modeled here is "the client disconnects once they have received the response body" - that means, `disconnected` should be set in `send_gen` once the last `http.response.body` has been sent. (Whereas `receive` blocks when it has no available message to send)

```suggestion
    async def recv_gen() -> AsyncGenerator[Message, None]:
        yield {"type": "http.request"}
        await disconnected.await()
        while True:
            yield {"type": "http.disconnect"}

    async def send_gen() -> AsyncGenerator[None, Message]:
        msg = yield
        if msg["type"] == "http.response.body" and not msg.get("more_body", False):
            disconnected.set()
```